cloud-controller-manager/do/firewall_controller_test.go

```diff
@@ -0,0 +1,788 @@
+				return []godo.Firewall{}, newFakeNotFoundResponse(), nil
+			},
+			expectedGodoFirewallCreateResp: func(context.Context, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
+				return newFakeFirewall(), newFakeOKResponse(), nil
+			},
+		},
+		{
+			name:    "failing to update the firewall because of an unexpected error",
+			fwCache: newFakeFirewallCacheWithDiffInboundRules(),
+			firewallRequest: &godo.FirewallRequest{
+				Name:          testWorkerFWName,
+				InboundRules:  testInboundRules,
+				OutboundRules: testOutboundRules,
+				Tags:          testWorkerFWTags,
+			},
+			expectedGodoFirewallUpdateResp: func(context.Context, string, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
+				return nil, newFakeNotFoundResponse(), errors.New("unexpected error")
+			},
+			expectedGodoFirewallAddRulesResp: func(context.Context, string, *godo.FirewallRulesRequest) (*godo.Response, error) {
+				return newFakeNotOKResponse(), errors.New("unexpected error")
+			},
+			expectedGodoFirewallListResp: func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error) {
+				return []godo.Firewall{*newFakeFirewall()}, newFakeOKResponse(), nil
+			},
+			expectedGodoFirewallCreateResp: func(context.Context, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
+				return nil, newFakeNotOKResponse(), errors.New("unexpected error")
+			},
+			expectedError: errors.New("failed to create firewall"),
+		},
+		{
+			name:    "failing to get the firewall",
+			fwCache: newFakeFirewallCacheEmpty(),
+			firewallRequest: &godo.FirewallRequest{
+				Name:          testWorkerFWName,
+				InboundRules:  testInboundRules,
+				OutboundRules: testOutboundRules,
+				Tags:          testWorkerFWTags,
+			},
+			expectedGodoFirewallGetResp: func(context.Context, string) (*godo.Firewall, *godo.Response, error) {
+				return nil, newFakeNotOKResponse(), errors.New("unexpected error")
+			},
+			expectedGodoFirewallListResp: func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error) {
+				return []godo.Firewall{*newFakeFirewall()}, newFakeNotOKResponse(), errors.New("unexpected error")
+			},
+			expectedError: errors.New("failed to create firewall"),
+		},
+		{
+			name:    "return error when failing to add inbound rules on update request to firewall API",
+			fwCache: newFakeFirewallCacheWithDiffInboundRules(),
+			firewallRequest: &godo.FirewallRequest{
+				Name:          testWorkerFWName,
+				InboundRules:  testInboundRules,
+				OutboundRules: testOutboundRules,
+				Tags:          testWorkerFWTags,
+			},
+			expectedGodoFirewallUpdateResp: func(context.Context, string, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
+				return nil, newFakeNotOKResponse(), errors.New("failed to add inbound rules")
+			},
+			expectedGodoFirewallAddRulesResp: func(context.Context, string, *godo.FirewallRulesRequest) (*godo.Response, error) {
```

Comment:
We should not be invoking `AddRules` anymore. Could you please update this test case and double-check that the others are aligned with the current implementation?